src/System.Security.Cryptography.X509Certificates/tests/Cert.cs

```diff
@@ -0,0 +1,159 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Xunit;
+
+namespace System.Security.Cryptography.X509Certificates.Tests
+{
+    //
+    // Helper class centralizes all loading of PFX's. Loading PFX's is a problem because of the key on disk that it creates and gets left behind
+    // if the certificate isn't properly disposed. As a result:
+    //
+    //    - PFX loading needs to be synchronized. If two threads race to load the same PFX, they may accidentally end up pointing to the same
+    //      key container. This means the first one that diposes deletes the key out from the other.
+    //
+    //    - Properly disposing PFX's imported into a X509Certificate2Collection is a pain because X509Certificate2Collection doesn't implement IDisposable.
+    //      To make this easier, we wrap these in an ImportedCollection class that does implement IDisposable.
+    //
+    // Though PFX's are the ones that really need this, it's fine to use this for ordinary DER data as well. It's probably better to use
+    // a consistent pattern throughout.
+    //
+    internal static class Cert
+    {
+        //
+        // The New() methods have an overload for each X509Certificate2 constructor that takes data.
+        //
+        public static X509Certificate2 New(byte[] rawData)
+        {
+            lock (SyncObject)
```

Comment:
>So the same PFX being loaded twice loads to the same key name.

That’s my hypothesis too. Actually, if the PFX embedded name is already taken, Windows gens up a Guid and uses that as the name. (There’s a flag (PKCS12_ALLOW_OVERWRITE_KEY) that changes the behavior to “nah, just use the same name and overwrite the key”, but our code isn’t passing that.) But given that this is synchronizing on a file system, I could easily believe there’s still a small window. This isn’t something we can fix on our side, though.

Assuming that’s correct, it seems synchronizing PFX loads (of the same blob data at least) is the solution. It’s true that this code only synchronizes within a process. And unless we’re currently parallelizing the same test across processes, it seems what this PR is doing is reasonable mitigation for these intermittent failures.